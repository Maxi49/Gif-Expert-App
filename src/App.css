@tailwind base;
@tailwind components;
@tailwind utilities;

.bg-custom-gradient {
  background-image: linear-gradient(to right, theme('colors.gray.800'), theme('colors.gray.400'), theme('colors.white'));
}
body {
  @apply bg-neutral-950;
  @apply w-96;
  @apply flex flex-row justify-start w-full min-h-screen p-12 ;
  @apply flex justify-center
}