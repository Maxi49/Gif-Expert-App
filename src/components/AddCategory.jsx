import React,{ useState } from "react";
import PropTypes from "prop-types";

export function AddCategory( { onNewCategory } ) {
    const [ inputValue,setInputValue] = useState('');
    {//! Creamos un useState con una variable de string vacia y una funcion para la misma para ir manejando el estado del input */
    }
    const onInputChange = ( event ) => {
        //console.log(event.target.value)
        setInputValue(event.target.value)
        //!Agarramos el value del input y le decimos que cuando cambie el valor de este tambien cambie el valor del inputvalue  
        //se puede desestructurar el target directamente
    }
    const handleSubmit = ( event ) => {
        event.preventDefault()
              //prevent del refresh
        
        //setCategories(categories => [inputValue,...categories]);
        if(inputValue.trim().length <= 1) return;
        onNewCategory(inputValue.trim());
        //le pasamos el value al prop
        setInputValue(''); 
    }
    return (
        <form onSubmit={ handleSubmit } className="flex justify-center" >
            <input className="px-4 py-2 mt-8 text-lg text-black bg-white border rounded shadow-xl outline-none w-60 md:w-full border-b-slate-950 " type="text" placeholder="Buscar gifs ej:Minecraft"
                value={inputValue /*Es el valor inicial del input */}
                onChange={ onInputChange /* cuando mandemos el form se 
            activa la funcion. Podemos mandar solo la referencia y esto ya recibe el value automaticamente */ } 

            />
    
        </form> 
    );
}

AddCategory.propTypes = {
    
};

export default AddCategory;
