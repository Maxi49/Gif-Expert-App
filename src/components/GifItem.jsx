import React from "react";

import PropTypes from 'prop-types'; 

export const GifItem = ({ title,url }) => {

return (
    <section className="p-5 md:p-2 lg:p-4 " >
        <li className="grid items-center content-center justify-center w-full p-4 overflow-hidden shadow-md rounded-2xl bg-neutral-900-700 h-96 md:h-96 shadow-slate-400 " >
            <img className="object-cover w-full h-72" src={url.url} alt={title} />
            <p className="text-zinc-300" >{title}</p>
        </li>
    </section>
)};

GifItem.propTypes = {
    title : PropTypes.string.isRequired,
    url : PropTypes.string.isRequired
}

