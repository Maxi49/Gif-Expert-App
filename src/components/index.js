export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem'

/*Basicamente al hacer un archivo index dentro de una carpeta y exportar todo ahi, ya con el simple echo de exportar la carpeta a algun otro archivo, js en si va a buscar el index de la carpeta con todo exportado y se pueden desestructurar todos los elementos desde ese import*/