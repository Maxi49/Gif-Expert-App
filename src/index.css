body {
    padding: 40px;
}

.btn{
    margin: 5px;
}