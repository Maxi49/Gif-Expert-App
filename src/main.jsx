import React from 'react'
import ReactDOM from 'react-dom/client'

import './index.css'
import { CounterApp } from './01-useState/CounterApp.jsx'
import { CounterHook } from './01-useState/CounterHook.jsx'
import { SimpleForm } from './02-useEffect/SimpleForm'
import RenderComponents from './RenderComponents'




ReactDOM.createRoot(document.getElementById('root')).render(
  <React.StrictMode>
    <RenderComponents/>
  </React.StrictMode>,
)
