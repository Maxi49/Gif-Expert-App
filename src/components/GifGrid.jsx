import PropTypes from "prop-types";
import { useEffect, useState } from "react";
import { GifItem } from "./GifItem";
import { useFetchGifs } from "../hooks/useFetchGifs";


function GifGrid({ category }) {
    const { images,isLoading } = useFetchGifs( category );
    //! nunca poner la creacion de una funcion dentro de un componente
    return (
    <section className="flex flex-col justify-center m-0 mb-16 list-none " >
        <h3 className="flex justify-center m-16 text-5xl text-gray-200" >{category}</h3>
        {
            isLoading && ( <h2>Loading...</h2> )
        }
        <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3" >
                { images.map(image => (
                    <GifItem key={ image.id }
                        { ...image }
                        //!spread de propiedades 
                    />
                ))}
        </div>
    </section>
    );
}

GifGrid.propTypes = {};

export default GifGrid;
