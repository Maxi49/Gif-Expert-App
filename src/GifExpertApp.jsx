import React, { useState, useRef } from "react";
import { AddCategory } from "./components";
import GifGrid from "./components/GifGrid";

export default function GifExpertApp() {
  const [categories, setCategories] = useState([]);

  const onAddCategory = (newCategory) => {
    const categoryExists = categories.find(
      (category) => category.toLowerCase() === newCategory.toLowerCase()
    );
    if (categoryExists) {
      return;
    }
    //pasamos newCategory como argumento de la funcion
    setCategories([ newCategory,...categories ]);
    //y la pasamos a nuestra funcion para cambiar el estado de la lista con el event enviado desde AddCategory
  };
  //la funcion tambine almacena el valor de la variable
  return (
    <>
      <section className="flex flex-col justify-center" >
        <h1 className="flex justify-center m-4 text-4xl sm:text-7xl text-neutral-200">Gif Expert App</h1>

        <AddCategory
          onNewCategory={onAddCategory}
          //mandamos una prop onNewCategory que manda la funcion onAddCategory
          //le llega el value mediante un event al prop y lo recibimos como newCategory
          //setCategories = { setCategories }
        />
          {categories.map((category) => (
            <GifGrid key={ category } 
            category = { category } />
          ))}
      </section>
    </>
  );
}
