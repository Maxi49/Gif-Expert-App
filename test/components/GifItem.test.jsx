import { render,screen } from "@testing-library/react"
import { GifItem } from "../../src/components"


describe('Test for GifItems', () => { 
    const title = 'Hola';
    const url = 'https://hola.com/hola.jpg'

    test('should match with the snapshot', () => { 
        const { conatiner } = render( <GifItem title={title} url={url} /> )
        expect( conatiner ).toMatchSnapshot()
    })

    test('should show the image with the indicated url',() => {
        render( <GifItem title={title} url={{url: url}}/> )
        const { src,alt } = screen.getByRole('img');

        expect(src).toBe(url);
        expect(alt).toBe(title)
            
    })

    test('should show the title in the component', () => { 

        render( <GifItem title={title} url={{url: url}}/> )

        expect( screen.getByText(title) ).toBeTruthy();
    })
})