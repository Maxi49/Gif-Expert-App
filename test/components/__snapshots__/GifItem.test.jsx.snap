// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for GifItems should match with the snapshot 1`] = `undefined`;
